how-it-works/1_protocol/03_message-routing.subpage.md

```diff
@@ -7,8 +7,128 @@ slug: message-routing
+It is important to note that this local inter-canister messaging is *not synchronous* message invocation as you might be used from EVM-based blockchains.
+Rather, local messages are put into the input queue of the target canister and are scheduled for execution asynchronously.
+This is the standard inter-canister messaging semantics known for the Internet Computer.
+
+### Inter-Subnet Inter-Canister Messaging
+
+Remote inter-canister messages, that is, such sent to canisters on other subnets, are handled differently.
+A remote message is placed into the outgoing *subnet queue* for the target subnet once it results from the execution of a canister message.
+The xnet messages in the queue are certified (signed) using a Merkle-tree-style data representation, at the end of the round by the subnet using BLS threshold cryptography as part of the per-round state certification.
+That is, every message in the outgoing queue is certified by the originating subnet.
+Replicas on the receiving subnet obtain the xnet message during block making (part of consensus), validate the threshold signature, and include a valid xnet message in a consensus block.
+Thanks to using a Merkle-tree-like xnet queue, parts of the queue can be consumed in a round by the receiving subnets and signatures can still be validated.
+
+## State Certification
+
+The replicated state of a subnet comprises all the relevant information required for the operation of the subnet:
+* Items certified per round:
+  * Responses to ingress messages
+  * Xnet messages to be sent to other subnets
+  * Canister metadata (module hashes, *certified variables*)
+* Items certified per checkpoint:
+  * All canister Wasm bytecode
+  * All canister state (heap and stable memory)
 
+The message routing layer performs a certification of parts of the replicated state after each round (the first three items above) and certification of the full replicated state (including also the remaining two items above) at every checkpoint, which is typically every multiple-hundred rounds.
+Certification is always done using BLS threshold signatures computed collectively by the subnet, thus certifications are computed by the subnet as a whole in a decentralized manner.
+
+### Per-Round Certification
+
+At the end of a round, i.e., when all messages have been executed or the cycles limit for the round has been reached (to ensure rounds cannot take arbitrarily long), the message routing layer performs a certification of parts of the replicated state.
+A BLS threshold signature is computed to certify the part of the state tree containing
+* Responses to ingress messages,
+* Xnet messages to be sent to other subnets, and
+* Canister metadata (module hashes, *certified variables*).
+
+The responses to ingress messages are referred to as ingress history.
+The certified responses can be read and validated against the subnet's public key by users as the response to their ingress messages.
```

Comment:
Should we also mention that each subnets public key is, in turn, certified as part of the NNS certification tree, which allows to verify things just based on the NNS public key?

Reply:
```suggestion
The certified responses can be read and validated against the subnet's public key by users as the response to their ingress messages. Each of the public keys of the individual subnets are, in turn, certified by the NNS using the same mechanism. This means that one can verify that certified responses indeed come from the IC only using the public key of the NNS.
```